[ao] Updated ModelReportVisualizer per-channel line plot

vspenubarthi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #82918
* #82917

Summary: Before, the line plot for the ModelReportVisualizer used to
plot a different line for each channel. However, for models that have a
lot of channels, this can get really hard to read and parse and doesn't
provide much valuable information.

Now, we just have a single value per module that is the average of the
500 channels.

We also considered plotting 3 lines (a min line, a max line, and an
average line) but the issue was that large outliers could result in one
of the lines completely messing up the scale and the other two not being
visible. As a result, it made sense to do an average and let the user
use the report data to generate the other two if they wished to do so.

This was tested visually in a ipynb notebook

Test Plan: Tested visually in a ipynb notebook

Reviewers:

Subscribers:

Tasks:

Tags:

vspenubarthi committed:
[ao] Updated ModelReportVisualizer per-channel line plot

Summary: Before, the line plot for the ModelReportVisualizer used to
plot a different line for each channel. However, for models that have a
lot of channels, this can get really hard to read and parse and doesn't
provide much valuable information.

Now, we just have a single value per module that is the average of the
500 channels.

We also considered plotting 3 lines (a min line, a max line, and an
average line) but the issue was that large outliers could result in one
of the lines completely messing up the scale and the other two not being
visible. As a result, it made sense to do an average and let the user
use the report data to generate the other two if they wished to do so.

This was tested visually in a ipynb notebook

Test Plan: Tested visually in a ipynb notebook

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/fx/_model_report/model_report_visualizer.py:
```diff
@@ -568,10 +567,14 @@ def generate_plot_visualization(self, feature_filter: str, module_fqn_filter: st
-            # plot a seperate line for each channel
-            for index, channel_info in enumerate(y_data):
-                ax.plot(x_data, channel_info, label="Channel {}".format(index))
+            # plot a single line that is average of the channel values
```
how is average for all channels useful to user?

vspenubarthi replied:
It will give them a general trend and if we want to show just one line of information, i felt this would be the most useful of having an average value be used.

The issue with showing all the different channels is that when you have 500+ channels, you really can't see anything cuz you have 500 lines on the plot.

Maybe for some types of information, this doesn't help as much, but for things like per_channel_weight_max or stats like that, getting an average max val across channels for each module and seeing that progression through the model can definately be useful in my opinion

jerryzh168 replied:
oh I mean when do they need to see this information, maybe can you document this any where? (as a non related example, when a user is trying to understand why we choose static quant v.s. dynamic quant, they might want to see some metric A)

jerryzh168 replied:
or in your answer, "getting an average max val across channels for each module and seeing that progression through the model can definately be useful in my opinion" can you talk in more detail about how exactly that is useful?

vspenubarthi replied:
well in the answer i provided, suppose users are trying to figure out why a set of input activations was classified as a non-stationary distribution or where the stationarity changes in a model. They can plot the average max and average min vals across the modules and that would give them a better sense of what is happening because the size of the batch ranges and epoch ranges per channel are what determine stationarity. 

Looking at the plots, they may realize that for a specific module C, the average input activation value max across channel sharply decreased across channels and the min sharply increased, leading to a lower stationarity (because batch range decrease -> stationarity decrease).

vspenubarthi replied:
Updated get_plot_visualization() method description to highlight reasoning a bit more.

vspenubarthi committed:
Update on "[ao] Updated ModelReportVisualizer per-channel line plot"


Summary: Before, the line plot for the ModelReportVisualizer used to
plot a different line for each channel. However, for models that have a
lot of channels, this can get really hard to read and parse and doesn't
provide much valuable information.

Now, we just have a single value per module that is the average of the
500 channels.

We also considered plotting 3 lines (a min line, a max line, and an
average line) but the issue was that large outliers could result in one
of the lines completely messing up the scale and the other two not being
visible. As a result, it made sense to do an average and let the user
use the report data to generate the other two if they wished to do so.

This was tested visually in a ipynb notebook

Test Plan: Tested visually in a ipynb notebook

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 reviewed (approved):
looks good, I feel there is a missing connection between how these metrics connects to the recommendations, are we expecting users to know all the details of how we are detecting stationary/non-stationary distributions and how we are giving recommendations here?

vspenubarthi commented:
@pytorchbot merge -g
